packages/trace-viewer/src/ui/modelUtil.ts

```diff
@@ -74,6 +75,41 @@ function indexModel(context: ContextEntry) {
     (event as any)[contextSymbol] = context;
 }
 
+function dedupeActions(actions: ActionTraceEvent[]) {
+  const callActions = actions.filter(a => a.metadata.id.startsWith('call@'));
+  const expectActions = actions.filter(a => a.metadata.id.startsWith('expect@'));
+
+  // Call startTime/endTime are server-side times.
+  // Expect startTime/endTime are client-side times.
+  // If there are call times, adjust expect startTime/endTime to align with callTime.
+  if (callActions.length && expectActions.length) {
+    const offet = callActions[0].metadata.startTime - callActions[0].metadata.wallTime!;
+    for (const expectAction of expectActions) {
+      const duration = expectAction.metadata.endTime - expectAction.metadata.startTime;
+      expectAction.metadata.startTime = expectAction.metadata.wallTime! + offet;
+      expectAction.metadata.endTime = expectAction.metadata.startTime + duration;
+    }
+  }
+  const callActionsByTime = new Map<number, ActionTraceEvent>();
+  for (const action of callActions)
+    callActionsByTime.set(action.metadata.wallTime!, action);
```

Comment:
What if there are two call actions at the same time?

Reply:
Hm, I ended up matching by time based on your recommendation.